src/Init/Data/Order/Factories.lean

```diff
@@ -0,0 +1,279 @@
+    IsLinearPreorder α where
+  toIsPreorder := .of_le
+  le_total := by simpa [LawfulOrderLE.le_iff] using le_total.total
+
+/--
+If an `OrderData α` instance is compatible with an `LE α` instance that is reflexive, antisymmetric
+and transitive, then the `OrderData α` instance is a partial order.
+-/
+public theorem IsPartialOrder.of_le {α : Type u} [OrderData α] [LE α] [LawfulOrderLE α]
+    (le_refl : Std.Refl (α := α) (· ≤ ·) := by exact inferInstance)
+    (le_antisymm : Std.Antisymm (α := α) (· ≤ ·) := by exact inferInstance)
+    (le_trans : Trans (α := α) (· ≤ ·) (· ≤ ·) (· ≤ ·) := by exact inferInstance) :
+    IsPartialOrder α where
+  toIsPreorder := .of_le
+  le_antisymm := by simpa [LawfulOrderLE.le_iff] using le_antisymm.antisymm
+
+/--
+If an `OrderData α` instance is compatible with an `LE α` instance that is antisymmetric,
+transitive and total, then the `OrderData α` instance is a linear order.
+-/
+public theorem IsLinearOrder.of_le {α : Type u} [OrderData α] [LE α] [LawfulOrderLE α]
+    (le_antisymm : Std.Antisymm (α := α) (· ≤ ·) := by exact inferInstance)
+    (le_trans : Trans (α := α) (· ≤ ·) (· ≤ ·) (· ≤ ·) := by exact inferInstance)
+    (le_total : Total (α := α) (· ≤ ·) := by exact inferInstance) :
+    IsLinearOrder α where
+  toIsLinearPreorder := .of_le
+  le_antisymm := by simpa [LawfulOrderLE.le_iff] using le_antisymm.antisymm
+
+/--
+Returns a `LawfulOrderLT α` instance given certain properties.
+
+If an `OrderData α` instance is compatible with an `LT α` instance, then this lemma derives
+a `LawfulOrderMin α` instance from two properties involving `LE α` and `Min α` instances.
+
+This convenience lemma combines `LawfulOrderInf.ofLE` and `LawfulOrderMin.ofLE`.
+-/
+public theorem LawfulOrderLT.ofLE {α : Type u} [OrderData α] [LT α] [LE α] [LawfulOrderLE α]
+    (lt_iff : ∀ a b : α, a < b ↔ a ≤ b ∧ ¬ b ≤ a) : LawfulOrderLT α where
+  lt_iff := by simpa [LawfulOrderLE.le_iff] using lt_iff
+
+/--
+If an `OrderData α` instance is compatible with an `LE α` instance, then this lemma characterizes
+in terms of `LE α` when a `Min α` instance "behaves like an infimum operator".
+-/
+public theorem LawfulOrderInf.ofLE {α : Type u} [OrderData α] [Min α] [LE α] [LawfulOrderLE α]
+    (le_min_iff : ∀ a b c : α, a ≤ Min.min b c ↔ a ≤ b ∧ a ≤ c) : LawfulOrderInf α where
+  le_min_iff := by simpa [LawfulOrderLE.le_iff] using le_min_iff
+
+/--
+Returns a `LawfulOrderMin α` instance given certain properties.
+
+If an `OrderData α` instance is compatible with an `LE α` instance, then this lemma derives
+a `LawfulOrderMin α` instance from two properties involving `LE α` and `Min α` instances.
+
+This convenience lemma combines `LawfulOrderInf.ofLE` and `LawfulOrderMin.ofLE`.
+-/
+public theorem LawfulOrderMin.ofLE {α : Type u} [OrderData α] [Min α] [LE α] [LawfulOrderLE α]
+    (le_min_iff : ∀ a b c : α, a ≤ Min.min b c ↔ a ≤ b ∧ a ≤ c)
+    (min_eq_or : ∀ a b : α, min a b = a ∨ min a b = b) : LawfulOrderMin α where
```

Comment:
Other convenient axiomatizations are possible here; am I correct in assuming that providing alternative constructors would be part of a future PR?

Reply:
Are you thinking of something roughly like "In a linear preorder where `Min` behaves like the left-leaning minimum, then there's a `LawfulOrderMin` instance? This is a good idea, I will provide such instances in a later PR.

If you are thinking of some other specific axiomatization, feel free to let me know here or in the review of said later PR.

Reply:
I was thinking of this, and also under the assumption of `min_eq_or`, I believe that `le_min_iff` is equivalent to `min a b ≤ a` and `min a b ≤ b`, which might be easier to provide.

Reply:
That's a very good idea. I even think your suggested axioms should be the default ones, so I'll move this lemma to a different name and provide the simpler version with `min a b ≤ a` and `min a b ≤ b` under the premium name instead.

Reply:
No, wait, this simpler version would require transitivity, which we currently do not assume here. Then I'd rather keep this lemma as-is and provide a separate one in a later PR.